client/js/widgets/dice.js

```diff
@@ -0,0 +1,394 @@
+class Dice extends Widget {
+  constructor(id) {
+    super(id);
+
+    this.facesElement = document.createElement('div');
+    this.facesElement.className = 'diceFaces';
+
+    this.domElement.appendChild(this.facesElement);
+
+    this.addDefaults({
+      width: 50,
+      height: 50,
+      typeClasses: 'widget dice',
+      classes: 'shaking',
+      clickable: true,
+      movable: false,
+      layer: 1,
+
+      borderRadius: '16%',
+      pipColor: 'black',
+      backgroundColor: 'white',
+      borderColor: null,
+
+      faces: [ 1, 2, 3, 4, 5, 6 ],
+      activeFace: 0,
+      rollCount: 0,
+
+      rollTime: 800,
+      swapTime: 500,
+
+      image: null,
+      text: null,
+      pips: null,
+      svgReplaces: null,
+      faceCSS: null,
+
+      pipSymbols: null,
+      shape3d: false
+    });
+  }
+
+  activeFace() {
+    return mod(Math.round(this.get('activeFace')), this.faceElements.length) || 0;
+  }
+
+  animate() {
+    if(!this.facesElement || !this.faceElements.length)
+      return;
+
+    const faceElements = this.faceElements;
+
+    const hash = this.rollHash ? this.rollHash : 0;
+    const fc = faceElements.length;
+    const af = this.activeFace();
+    const f0 = this.previousActiveFace;
+    let f1 = (f0 + (hash) % (fc - 1) + 1) % fc || 0;
+    if(f1 == f0)
+      f1 = (f1 + 1) % fc;
+    let f2 = (f1 + (hash >>> 14) % (fc - (f1 == af ? 1 : 2)) + 1) % fc || (f1 + 1) % fc;
+    if(f2 == af)
+      f2 = (f2 + 1) % fc;
+
+    this.facesElement.className = `diceFaces shake${hash >>> 29}`;
+
+    for(const face of faceElements)
+      face.classList.remove('animate1active', 'animate2active');
+    faceElements[f1].classList.add('animate1active');
+    faceElements[f2].classList.add('animate2active');
+  }
+
+  animateProperties() {
+    const rules = super.animateProperties();
+    const swap = this.get('swapTime');
+    const roll = this.get('rollTime');
+    rules.unshift(
+      { property: 'activeFace', duration: typeof swap == 'number'? swap : 1000},
+      { property: 'rollCount', duration: typeof roll == 'number'? roll + 50 : 1000},
+      { property: 'rollCount', className: 'animateBegin', duration: 50 }
+    );
+    return rules;
+  }
+
+  applyDeltaToDOM(delta) {
+    super.applyDeltaToDOM(delta);
+
+    if(delta.activeFace !== undefined || delta.faces !== undefined) {
+      let face = this.faces()[this.get('activeFace')];
+      if(face !== undefined)
+        this.applyDeltaToDOM(face);
```

Comment:
This can trigger createChildNodes() unnecessarily when only `activeFace` changes if the face changed to contains any of 'faces', 'shape3d', 'pipSymbols', 'faceCSS', 'image', 'text', 'pips', 'svgReplaces'.

Reply:
I think `"faces":[{"faces":null}]` may even put the client in an infinite loop.

Reply:
The way face linked properties currently work on dice, only properties not set in the widget are read from the face (except when building the child nodes). This means that `this.faces()this.get('activeFace')` returns the same value when running inside the recursive `this.applyDeltaToDOM(face)`, so setting either `faces` or `activeFace` in a face can crash the client in an infinite loop.

Reply:
How do we want dice face linked properties to work? The current code changes the default value of the property while the face is active (as well as changing the property when the face DOM element is built), so setting `faceCSS` on one face but not having a default `faceCSS` in the widget will apply the css only to that face until that face becomes the active face, then it will apply to all faces that do not have `faceCSS` set. When the face changes, the faces may not revert to no `faceCSS` because `this.applyDeltaToDOM(face)` may not trigger `createChildNodes()`. This means that the DOM state would be different for a newly connected client.

Reply:
Rather than recursively calling applyDeltaToDOM(face) can we just add the face properties to the delta currently being processed? We should also ensure that any property set by the prior face is in the delta.

Reply:
I thought my latest commit would fix the inconsistency issues. The recursive call is gone, so the infinite loop does not happen; but faces are not being reconstructed when changing away from a face that set properties used as the default on other faces.